app/commands/iteration/destroy.rb

```diff
@@ -7,14 +7,15 @@ class Destroy
     def call
       iteration.update!(deleted_at: Time.current)
 
-      # Unpublish the solution if this was the published iteration
-      solution.update!(published_iteration_id: nil, published_at: nil) if solution.published_iteration_id == iteration.id
-
-      # Uncomplete the solution if there are now no iterations
-      solution.update!(completed_at: nil) unless solution.iterations.not_deleted.exists?
+      if !solution.iterations.not_deleted.exists?
+        Solution::Uncomplete.(solution)
+      elsif solution.published_iteration_id == iteration.id
+        Solution::Unpublish.(solution)
+      end
```

Comment:
How about this to make it more readable. The second bit might run twice, but will just be a no-op the second time so it's not really painful.

```suggestion
      Solution::Uncomplete.(solution) unless solution.iterations.not_deleted.exists?
      Solution::Unpublish.(solution) if solution.published_iteration_id == iteration.id
```